# Exercise 4.7

# Table of Contents
1. Importing Data
2. Creating labels for products
3. Using .loc instead of if-statements
4. Creating loop for busiest days
5. Creating loop for hours of day

# 1. Importing Data

In [1]:
#importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
#defining path
path = r'C:\Users\jonsa\Documents\Instacart Basket Analysis'

In [3]:
ords_prods_merged = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data' , 'orders_products_merged.pkl'))

In [4]:
df = ords_prods_merged[:1000000]

# 2. Creating labels for products

In [5]:
#creating an if-statement to label if products are low, mid, or high range
def price_label(row):
    
    if row['prices'] <= 5:
        return 'Low range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid range product'
    elif row['prices'] > 15:
        return 'High range product'
    else: return np.nan

In [6]:
df['price_label'] = df.apply(price_label, axis=1)

C:\Users\jonsa\AppData\Local\Temp\ipykernel_9792\1527635097.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_label'] = df.apply(price_label, axis=1)


In [7]:
#counting values within each category
df['price_label'].value_counts()

Mid range product    756450
Low range product    243550
Name: price_label, dtype: int64

In [8]:
df['prices'].max()

14.8

**#Our max priced product is $14.80 meaning our high range product statement will be empty**

**#Creating a df.loc to save on memory instead of filtering only the first 1,000,000 rows**

# 3. Using .loc instead of if-statement

In [9]:
ords_prods_merged.loc[ords_prods_merged['prices'] >  15, 'price_range_loc'] = 'High range product'

In [10]:
ords_prods_merged.loc[(ords_prods_merged['prices'] <= 15) & (ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [11]:
ords_prods_merged.loc[ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [12]:
#Returning results from entire dataframe
ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High range product      417678
Name: price_range_loc, dtype: int64

In [13]:
#Identifying most popular days of the week. 0 = Saturday
ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

# 4. Creating loop for busiest days

In [14]:
#Creating a for-loop to create a new column 'busiest_day' to mark which days are busiest, least busy, and average
result = []

for value in ords_prods_merged['orders_day_of_week']:
    if value == 0:
        result.append('Busiest day')
    elif value == 4:
        result.append('Least busy day')
    else:
        result.append('Regular busy day')

In [15]:
ords_prods_merged['busiest_day'] = result

In [16]:
ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regular busy day    22416875
Busiest day          6204182
Least busy day       3783802
Name: busiest_day, dtype: int64

In [17]:
ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

**#Task**

In [18]:
#Creating a new for loop that identifies the two busiest and least busy days
result2 = []

for value in ords_prods_merged['orders_day_of_week']:
  if value in [0,1]:
    result2.append("Busiest days")
  elif value in [3,4]:
    result2.append("Slowest days")
  else:
    result2.append("Regularly busy")

In [19]:
ords_prods_merged['busiest_days'] = result2

In [20]:
#Printing the results of our if-loop
ords_prods_merged['busiest_days'].value_counts(dropna = False)

Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: busiest_days, dtype: int64

In [21]:
ords_prods_merged.columns

Index(['Unnamed: 0.1', 'Unnamed: 0_x', 'order_id', 'user_id', 'order_number',
       'orders_day_of_week', 'order_hour_of_day', 'days_since_previous_order',
       'only_order', 'product_id', 'add_to_cart_order', 'reordered', '_merge',
       'Unnamed: 0_y', 'product_name', 'aisle_id', 'department_id', 'prices',
       'price_range_loc', 'busiest_day', 'busiest_days'],
      dtype='object')

In [22]:
#Identifying which hours are the busiest and least busy
ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

# 5. Creating loop for hours of day

In [26]:
#Creating a for loop to label all values according to orders made during each hour
result3 = []

for value in ords_prods_merged['order_hour_of_day']:
  if value >= 10 and value <= 16:
    result3.append("Most orders")
  elif value <= 6:
    result3.append("Least orders")
  else:
    result3.append("Average orders")

In [27]:
ords_prods_merged['busiest_period_of_day'] = result3

In [28]:
#printing results of the for loop
ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       18663868
Average orders    12854170
Least orders        886821
Name: busiest_period_of_day, dtype: int64

In [29]:
ords_prods_merged.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'ords_prods_merged_2.pkl'))